README.md
# Remote Scala Eval

The remote Scala evaluator is a server based application that
allows remote evaluation of arbitrary Scala code.

# Run from sources

```bash
sbt "project evaluator-server" "run"
```

# Authentication

The remote Scala eval uses [JWT](https://jwt.io/) to encode / decode tokens.
The `secretKey` used for encoding/decoding is configurable as part of the service configuration in
`server/src/main/resources/application.conf`.

Please change `secretKey` by overriding it or providing the `EVAL_SECRET_KEY` env var.

```
eval.auth {
  secretKey = "secretKey"
  secretKey = ${?EVAL_SECRET_KEY}
}
```

## Generate an auth token

In order to generate an auth token you may use the scala console and invoke
the `org.scalaexercises.evaluator.auth#generateToken` like so

```
sbt "project evaluator-server" "console"

scala> import org.scalaexercises.evaluator.auth._

scala> generateToken("your identity")
res0: String = eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eW91ciBpZGVudGl0eQ.cfH43Wa7k_w1i0W2pQhV1k21t2JqER9lw5EpJcENRMI
```

Note `your identity` is exclusively to identify incoming requests for logging purposes.
The Scala evaluator will authorize any incoming request generated with the `secretKey`

# Request

Requests are sent in JSON format via HTTP POST and are authenticated via the `x-scala-eval-api-token` header.

# Sample Request

Given the token above a sample request may look like:

```bash
curl -X POST -H "Content-Type: application/json" -H "x-scala-eval-api-token: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eW91ciBpZGVudGl0eQ.cfH43Wa7k_w1i0W2pQhV1k21t2JqER9lw5EpJcENRMI" -d '{
   "resolvers":[
      "https://oss.sonatype.org/content/repositories/releases"
   ],
   "dependencies":[
      {
         "groupId":"org.typelevel",
         "artifactId":"cats-core_2.11",
         "version":"0.4.1"
      }
   ],
   "code":"{import cats._; Monad[Id].pure(42)}"
}
' "http://localhost:8080/eval"
```

## Headers

```
Content-Type : application/json
x-scala-eval-api-token : eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eW91ciBpZGVudGl0eQ.cfH43Wa7k_w1i0W2pQhV1k21t2JqER9lw5EpJcENRMI 
```

## Body

```json
{
   "resolvers":[
      "https://oss.sonatype.org/content/repositories/releases"
   ],
   "dependencies":[
      {
         "groupId":"org.typelevel",
         "artifactId":"cats-core_2.11",
         "version":"0.4.1"
      }
   ],
   "code":"{import cats._; Monad[Id].pure(42)}"
}
```

- `resolvers` : A list of resolvers where artifacts dependencies are hosted
- `dependencies` : A List of artifacts required to eval the code
- `code` : Some Scala Code

## Response

After compiling and attempting to evaluate the server will return a response payload with the following fields:

- `msg` : A message indicating the result of the compilation and evaluation. Note failing to compile still yields http status 200 as the purpose of the service is to output a result without judging if the input was correct or not.	
- `value` : The result of the evaluation or `null` if it didn't compile or an exception is thrown.
- `valueType` : The type of result or `null` if it didn't compile or an exception is thrown
- `compilationInfos` : A map of compilation severity errors and their associated message

For ilustration purposes here is a few Response Payload examples:

Successful compilation and Evaluation

```json
{
  "msg": "Ok",
  "value": "42",
  "valueType": "java.lang.Integer",
  "compilationInfos": {}
}
```

Compilation Failure

```json
{
  "msg": "Compilation Error",
  "value": null,
  "valueType": null,
  "compilationInfos": {
    "ERROR": [
      {
        "message": "value x is not a member of cats.Monad[cats.Id]",
        "pos": {
          "start": 165,
          "point": 165,
          "end": 165
        }
      }
    ]
  }
}
```

Evaluating code that may result in a thrown exception

```json
{
  "msg": "Runtime Error",
  "value": null,
  "valueType": null,
  "compilationInfos": {}
}
```


CHANGELOG.md
# Changelog

## [v0.1.3](https://github.com/scala-exercises/evaluator/tree/v0.1.3) (2020-08-10)

[Full Changelog](https://github.com/scala-exercises/evaluator/compare/v0.1.2...v0.1.3)

📈 **Dependency updates**

- Update sbt-mdoc to 2.2.4 [\#160](https://github.com/scala-exercises/evaluator/pull/160) ([47erbot](https://github.com/47erbot))
- Update http4s-blaze-server, http4s-circe, ... to 0.21.7 [\#159](https://github.com/scala-exercises/evaluator/pull/159) ([47erbot](https://github.com/47erbot))
- Update sbt-buildinfo to 0.10.0 [\#158](https://github.com/scala-exercises/evaluator/pull/158) ([47erbot](https://github.com/47erbot))
- Update scalacheck-1-14 to 3.2.1.0 [\#157](https://github.com/scala-exercises/evaluator/pull/157) ([47erbot](https://github.com/47erbot))
- Update scalatest to 3.2.1 [\#156](https://github.com/scala-exercises/evaluator/pull/156) ([47erbot](https://github.com/47erbot))
- Update sbt-scalafmt to 2.4.2 [\#155](https://github.com/scala-exercises/evaluator/pull/155) ([47erbot](https://github.com/47erbot))
- Update coursier, coursier-cache, ... to 2.0.0-RC6-24 [\#154](https://github.com/scala-exercises/evaluator/pull/154) ([47erbot](https://github.com/47erbot))
- Update scalafmt-core to 2.6.4 [\#153](https://github.com/scala-exercises/evaluator/pull/153) ([47erbot](https://github.com/47erbot))
- Update cats-effect to 2.1.4 [\#152](https://github.com/scala-exercises/evaluator/pull/152) ([scala-steward](https://github.com/scala-steward))
- Update scalafmt-core to 2.6.3 [\#151](https://github.com/scala-exercises/evaluator/pull/151) ([scala-steward](https://github.com/scala-steward))
- Update sbt-docker to 1.7.0 [\#150](https://github.com/scala-exercises/evaluator/pull/150) ([scala-steward](https://github.com/scala-steward))
- Update coursier, coursier-cache, ... to 2.0.0-RC6-23 [\#149](https://github.com/scala-exercises/evaluator/pull/149) ([scala-steward](https://github.com/scala-steward))
- Update sbt-native-packager to 1.7.4 [\#148](https://github.com/scala-exercises/evaluator/pull/148) ([scala-steward](https://github.com/scala-steward))
- Update coursier, coursier-cache, ... to 2.0.0-RC6-22 [\#147](https://github.com/scala-exercises/evaluator/pull/147) ([scala-steward](https://github.com/scala-steward))
- Update scalafmt-core to 2.6.2 [\#146](https://github.com/scala-exercises/evaluator/pull/146) ([scala-steward](https://github.com/scala-steward))
- Update sbt-docker to 1.6.0 [\#145](https://github.com/scala-exercises/evaluator/pull/145) ([scala-steward](https://github.com/scala-steward))

## [v0.1.2](https://github.com/scala-exercises/evaluator/tree/v0.1.2) (2020-06-29)

[Full Changelog](https://github.com/scala-exercises/evaluator/compare/v0.1.1...v0.1.2)

📈 **Dependency updates**

- Update sbt-mdoc to 2.2.3 [\#136](https://github.com/scala-exercises/evaluator/pull/136) ([scala-steward](https://github.com/scala-steward))
- Update http4s-blaze-server, http4s-circe, ... to 0.21.5 [\#135](https://github.com/scala-exercises/evaluator/pull/135) ([scala-steward](https://github.com/scala-steward))
- Update scalacheck-1-14 to 3.2.0.0 [\#131](https://github.com/scala-exercises/evaluator/pull/131) ([scala-steward](https://github.com/scala-steward))
- Update scalatest to 3.2.0 [\#130](https://github.com/scala-exercises/evaluator/pull/130) ([scala-steward](https://github.com/scala-steward))
- Update scalafmt-core to 2.5.3 [\#126](https://github.com/scala-exercises/evaluator/pull/126) ([scala-steward](https://github.com/scala-steward))
- Update sbt-mdoc to 2.2.2 [\#125](https://github.com/scala-exercises/evaluator/pull/125) ([scala-steward](https://github.com/scala-steward))
- Update sbt-native-packager to 1.7.3 [\#124](https://github.com/scala-exercises/evaluator/pull/124) ([scala-steward](https://github.com/scala-steward))

**Merged pull requests:**

- Updates build [\#140](https://github.com/scala-exercises/evaluator/pull/140) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Update scalafmt-core to 2.6.1 [\#133](https://github.com/scala-exercises/evaluator/pull/133) ([BenFradet](https://github.com/BenFradet))
- Update scalafmt-core to 2.6.0 [\#128](https://github.com/scala-exercises/evaluator/pull/128) ([BenFradet](https://github.com/BenFradet))

## [v0.1.1](https://github.com/scala-exercises/evaluator/tree/v0.1.1) (2020-06-09)

[Full Changelog](https://github.com/scala-exercises/evaluator/compare/v0.1.0...v0.1.1)

📈 **Dependency updates**

- Update sbt to 1.3.11 [\#119](https://github.com/scala-exercises/evaluator/pull/119) ([scala-steward](https://github.com/scala-steward))
- Update scalatest to 3.1.2 [\#117](https://github.com/scala-exercises/evaluator/pull/117) ([scala-steward](https://github.com/scala-steward))
- Update sbt-native-packager to 1.7.1 [\#113](https://github.com/scala-exercises/evaluator/pull/113) ([scala-steward](https://github.com/scala-steward))
- Update joda-time to 2.10.6 [\#110](https://github.com/scala-exercises/evaluator/pull/110) ([scala-steward](https://github.com/scala-steward))

**Merged pull requests:**

- Prepare repository for next `.github` release and SBT build improvements [\#122](https://github.com/scala-exercises/evaluator/pull/122) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Update scalafmt-core to 2.5.1 [\#116](https://github.com/scala-exercises/evaluator/pull/116) ([BenFradet](https://github.com/BenFradet))
- Adds scmInfo settings [\#112](https://github.com/scala-exercises/evaluator/pull/112) ([juanpedromoreno](https://github.com/juanpedromoreno))

## [v0.1.0](https://github.com/scala-exercises/evaluator/tree/v0.1.0) (2020-04-23)

[Full Changelog](https://github.com/scala-exercises/evaluator/compare/da6604a17a43c748f59436eb533986c6b041529f...v0.1.0)

🚀 **Features**

- Upgrade Cats and Circe Version [\#34](https://github.com/scala-exercises/evaluator/issues/34)

🐛 **Bug Fixes**

- Fix macroparadise's inclusion as a compiler plugin [\#55](https://github.com/scala-exercises/evaluator/issues/55)

**Closed issues:**

project/ProjectPlugin.scala
import sbt.Keys._
import sbt._
import sbtassembly.AssemblyPlugin.autoImport.assembly
import sbtbuildinfo.BuildInfoKey
import sbtbuildinfo.BuildInfoPlugin.autoImport._
import sbtdocker.DockerPlugin.autoImport._
import com.alejandrohdezma.sbt.github.SbtGithubPlugin

object ProjectPlugin extends AutoPlugin {

  override def trigger: PluginTrigger = allRequirements

  override def requires: Plugins = plugins.JvmPlugin && SbtGithubPlugin

  object autoImport {

    object V {
      lazy val cats                = "2.1.1"
      lazy val catsEffect          = "2.1.4"
      lazy val http4s              = "0.21.7"
      lazy val circe               = "0.13.0"
      lazy val log4s               = "1.7.0"
      lazy val scalatest           = "3.2.1"
      lazy val scalatestplusScheck = "3.2.1.0"
      lazy val jodaTime            = "2.10.6"
      lazy val slf4j               = "1.7.30"
      lazy val jwtCore             = "4.3.0"
      lazy val coursier            = "2.0.0-RC6-24"
      lazy val config              = "1.4.0"
      lazy val scala               = "2.13.3"
    }

    lazy val dockerSettings = Seq(
      docker := (docker dependsOn assembly).value,
      dockerfile in docker := {

        val artifact: File     = assembly.value
        val artifactTargetPath = artifact.name

        sbtdocker.immutable.Dockerfile.empty
          .from("ubuntu:latest")
          .run("apt-get", "update")
          .run("apt-get", "install", "-y", "openjdk-8-jdk")
          .run("useradd", "-m", "evaluator")
          .user("evaluator")
          .add(artifact, artifactTargetPath)
          .cmdRaw(
            s"java -Dhttp.port=$$PORT -Deval.auth.secretKey=$$EVAL_SECRET_KEY -jar $artifactTargetPath"
          )
      },
      imageNames in docker := Seq(
        ImageName(repository =
          s"registry.heroku.com/${sys.props.getOrElse("evaluator.heroku.name", "scala-evaluator")}/web"
        )
      )
    )

    lazy val serverScalaMacroDependencies: Seq[Setting[_]] = {
      Seq(
        libraryDependencies += "org.scala-lang" % "scala-compiler" % scalaVersion.value,
        libraryDependencies += "org.scala-lang" % "scala-reflect"  % scalaVersion.value
      )
    }

    lazy val serverHttpDependencies = Seq(
      libraryDependencies ++= Seq(
        "org.typelevel"     %% "cats-core"             % V.cats,
        "org.typelevel"     %% "cats-effect"           % V.catsEffect,
        "io.circe"          %% "circe-core"            % V.circe,
        "io.circe"          %% "circe-generic"         % V.circe,
        "org.slf4j"          % "slf4j-simple"          % V.slf4j,
        "org.http4s"        %% "http4s-dsl"            % V.http4s,
        "org.http4s"        %% "http4s-blaze-server"   % V.http4s,
        "org.http4s"        %% "http4s-circe"          % V.http4s,
        "io.get-coursier"   %% "coursier"              % V.coursier,
        "io.get-coursier"   %% "coursier-cache"        % V.coursier,
        "com.typesafe"       % "config"                % V.config,
        "com.pauldijou"     %% "jwt-core"              % V.jwtCore,
        "io.get-coursier"   %% "coursier-cats-interop" % V.coursier,
        "org.scalatest"     %% "scalatest"             % V.scalatest,
        "org.scalatestplus" %% "scalacheck-1-14"       % V.scalatestplusScheck,
        "joda-time"          % "joda-time"             % V.jodaTime
      )
    )

    lazy val buildInfoSettings = Seq(
      buildInfoKeys := Seq[BuildInfoKey](name, version, scalaVersion, sbtVersion),
      buildInfoPackage := "org.scalaexercises.evaluator"
    )

  }

  override def projectSettings: Seq[Def.Setting[_]] =
    Seq(
      scmInfo := Some(
        ScmInfo(
          url("https://github.com/scala-exercises/evaluator"),
          "scm:git:https://github.com/scala-exercises/evaluator.git",
          Some("scm:git:git@github.com:scala-exercises/evaluator.git")
        )
      ),
      scalacOptions ~= (_ filterNot (_ == "-Xfuture")),
      scalacOptions += "-Ymacro-annotations",
      javacOptions ++= Seq("-encoding", "UTF-8", "-Xlint:-options"),
      parallelExecution in Test := false,
      cancelable in Global := true
    )
}
